hy/importer.py

```diff
@@ -68,34 +71,77 @@ def import_file_to_ast(fpath, module_name):
+    """Import Hy source from fpath and put it into a Python module.
 
-    Returns the module."""
-    try:
-        _ast = import_file_to_ast(fpath, module_name)
-        mod = imp.new_module(module_name)
-        mod.__file__ = fpath
-        eval(ast_compile(_ast, fpath, "exec"), mod.__dict__)
-    except (HyTypeError, LexException) as e:
-        if e.source is None:
-            with open(fpath, 'rt') as fp:
-                e.source = fp.read()
-            e.filename = fpath
-        raise
-    except Exception:
-        sys.modules.pop(module_name, None)
-        raise
-    return mod
+    If there's an up-to-date byte-compiled version of this module, load that
+    instead. Otherwise, byte-compile the module once we're done loading it, if
+    we can.
 
+    Return the module."""
 
-def import_file_to_globals(env, module_name, fpath):
-    """ Import content from fpath and puts it into the dict provided
-    (e.g., for use in a REPL)
-    """
-    mod = import_file_to_module(module_name, fpath)
-    for k, v in mod.__dict__.items():
-        env[k] = v
+    module = None
+
+    bytecode_path = get_bytecode_path(fpath)
+    try:
+        source_mtime = int(os.stat(fpath).st_mtime)
+        with open(bytecode_path, 'rb') as bc_f:
+            # To get the bytecode file's internal timestamp, take the 4 bytes
+            # after the first 4 bytes and interpret them as a little-endian
+            # 32-bit integer.
+            bytecode_mtime = struct.unpack('<i', bc_f.read(8)[4:])[0]
+    except (IOError, OSError):
+        pass
+    else:
+        if bytecode_mtime >= source_mtime:
+            # It's a cache hit. Load the byte-compiled version.
+            if PY3:
+                # As of Python 3.6, imp.load_compiled still exists, but it's
+                # deprecated. So let's use SourcelessFileLoader instead.
+                from importlib.machinery import SourcelessFileLoader
+                module = (SourcelessFileLoader(module_name, bytecode_path).
+                          load_module(module_name))
+            else:
+                module = imp.load_compiled(module_name, bytecode_path)
+
+    if not module:
+        # It's a cache miss, so load from source.
+        sys.modules[module_name] = None
+        try:
+            _ast = import_file_to_ast(fpath, module_name)
+            module = imp.new_module(module_name)
+            module.__file__ = fpath
+            code = ast_compile(_ast, fpath, "exec")
+            try:
+                write_code_as_pyc(fpath, code)
+            except IOError:
+                # We failed to save the bytecode, probably because of a
+                # permissions issue. The user only asked to import the file, so
+                # don't bug them about it.
+                pass
+            eval(code, module.__dict__)
+        except (HyTypeError, LexException) as e:
+            if e.source is None:
+                with open(fpath, 'rt') as fp:
```

Comment:
Maybe also wrap this in a `try...except`? In the theoretical case where a source file disappears before the error is shown (don't laugh; I've accidentally un-plugged my external HDD and run into trouble before), this would outright crash, while maybe just printing a warning or just doing nothing would be a bit better.